File: 프로그래머스/1/42576. 완주하지 못한 선수/완주하지 못한 선수.cpp
#include <string>
#include <vector>
#include <map>

using namespace std;

string solution(vector<string> participant, vector<string> completion) {
    string answer = "";
    
    map<string, int> p;
    
    for (int i = 0; i < participant.size(); i++) {
            p[participant[i]]++;
    }
    for (int i = 0; i < completion.size(); i++) {
        string name = completion[i];
        if (p[name] > 1)
            p[name]--;
        else
            p.erase(name);
    }
    
    answer = p.begin() -> first;
    return answer;
}

File: 프로그래머스/1/42576. 완주하지 못한 선수/완주하지 못한 선수.java
import java.util.HashMap;
import java.util.ArrayList;
import java.util.List;

class Solution {
    public String solution(String[] participant, String[] completion) {
        String answer = "";

        HashMap<String, List<Integer>> map = new HashMap<>();
        for (int i = 0; i < completion.length; i++) {
            map.computeIfAbsent(completion[i], k -> new ArrayList<>()).add(i);
        }
        
        for (int i = 0; i < participant.length; i++) {
            if (map.containsKey(participant[i])) {
                List<Integer> list = map.get(participant[i]);
                if (list.isEmpty())
                    return participant[i];
                list.remove(list.size() - 1);
            }
            else {
                answer = participant[i];
                break;
            }
        }
        return answer;
    }
}